### PR TITLE
[WebPubSub] Add client constructors with AzureIdentity.TokenCredential to support AAD

### PR DESCRIPTION
**Pull request by terencefan**
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/main/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

**Comment by vicancy** [comment]
Hi @KrzysztofCwalina , could you help review the code? Appreciate your time!

**Review by JialinXin — APPROVED** [review]

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_extensions.cs [hunk 1]
```diff
@@ -43,7 +41,49 @@ public WebPubSubServiceClient(Uri endpoint, string hub, AzureKeyCredential crede
...
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of WebPubSubServiceClient.
+        /// </summary>
+        /// <param name="endpoint"></param>
+        /// <param name="hub"></param>
+        /// <param name="credential"></param>
+        private WebPubSubServiceClient(Uri endpoint, string hub, ITokenCredential credential)
```
[the remark this answers is not in the document]
Comment (terencefan) [reply sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_extensions.cs]: @pakrym I've updated the PR, could you review it again?

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/Azure.Messaging.WebPubSub.csproj [hunk 2]
```diff
@@ -12,6 +12,7 @@
   <ItemGroup>
     <PackageReference Include="Azure.Core" />
     <PackageReference Include="Azure.Core.Experimental" />
+    <PackageReference Include="Azure.Identity" />
```
[the remark this answers is not in the document]
Comment (terencefan) [reply sdk/webpubsub/Azure.Messaging.WebPubSub/src/Azure.Messaging.WebPubSub.csproj]: Sure, I have removed it from the PR.

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_helpers.cs [hunk 3]
```diff
@@ -145,7 +158,117 @@ internal static (Uri Endpoint, AzureKeyCredential Credential) ParseConnectionStr
...
+            {
+                return (uriBuilder.Uri, new AccessKeyTokenCredential(accessKey));
+            }
+            else
+            {
+                throw new ArgumentException($"Required property not found in connection string: {AccessKeyPropertyName}.");
+            }
+        }
+
+        internal interface ITokenCredential
```
[the remark this answers is not in the document]
Comment (terencefan) [reply sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_helpers.cs]: I found the naming is cofusing so I renamed it to `IHubEndpoint` and now we have `AccessKeyHubEndpoint` and `AzureIdentityHubEndpoint` for local auth and aad auth method respectively.

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_helpers.cs [hunk 4]
```diff
@@ -145,7 +158,117 @@ internal static (Uri Endpoint, AzureKeyCredential Credential) ParseConnectionStr
                 uriBuilder.Port = port.Value;
             }
 
-            return (uriBuilder.Uri, new AzureKeyCredential(accessKey));
+            if (dict.TryGetValue(AuthTypePropertyName, out var authType) && authType == "aad")
```
[the remark this answers is not in the document]
Comment (terencefan) [reply sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_helpers.cs]: All right, I have removed the parser from the PR.

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_extensions.cs [hunk 5]
```diff
@@ -70,36 +118,11 @@ public WebPubSubServiceClient(Uri endpoint, string hub, AzureKeyCredential crede
             Pipeline = HttpPipelineBuilder.Build(
                 options,
                 perCallPolicies: perCallPolicies,
-                perRetryPolicies: new HttpPipelinePolicy[] { new WebPubSubAuthenticationPolicy(credential) },
+                perRetryPolicies: new HttpPipelinePolicy[] { new WebPubSubAuthenticationPolicy(endpoint) },
```
Comment (pakrym) [inline sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_extensions.cs:121]: I recommend using `BearerTokenAuthenticationPolicy` for `TokenCredential`. It supports many features like token caching and background refresh.

**Comment by Mohit-Chakraborty** [comment]
Not sure why there were 2 separate efforts to update the SDK. Sorry for the confusion.

The SDK team has [added this support](https://github.com/Azure/azure-sdk-for-net/commit/953cff565fc83a7b54348db8308851a10958cb74#diff-65b27a70d3f718bdbaedefe746428f557793ec7c5f5328e7ed089bb89698d41a). Please take a look at the change and create a PR to make further changes if needed.

**Comment by vicancy** [comment]
I believe this PR is no longer needed. All the features should already be covered by @Mohit-Chakraborty 's PR. I am closing this one.

**Comment by vicancy** [comment]
Previously we thought the service team would be responsible for update the `GenerateClientAccessUri` to use the REST API with AAD credential. That's why we sent out this PR proposing the related changes. 

Thank you @Mohit-Chakraborty for your decent work supporting this!

**Pull request closed without merging**